### PR TITLE
Presigned URLs [DEVINFRA-933]

### PR DESCRIPTION
**Pull request by joelynch**
Added some functions for generating pre-signed urls for S3 Get, Put, Delete and Multipart Put.  Since it can be tricky to use pre-signed urls properly, which I found out during testing, I've also added in helpers for using pre-signed urls using the `reqwest` library.

TODO
- [ ] handle transparent compression for the multipart put
- [ ] test transparent compression
- [x] fix openssl missing error on CICD
- [ ] Add `presign` command to CLI

### CODE REVIEW

**Review by gborough — APPROVED** [review]

File: crates/esthri/src/presign/upload.rs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+use std::{os::unix::prelude::MetadataExt, path::Path, time::Duration};
```
Comment (adrian-kong) [inline crates/esthri/src/presign/upload.rs:13]:
```suggestion
use std::{path::Path, time::Duration};
```
think this unix specific - can delete this if we use len() - (*not actually sure the difference tho the docs sounds like they do exactly the same thing*)

File: crates/esthri/src/presign/upload.rs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+/// Helper to download a file using a presigned URL, setting appropriate
+/// headers.
+pub async fn upload_file_presigned(
+    client: &Client,
+    presigned_url: &str,
+    filepath: &Path,
+    opts: EsthriPutOptParams,
+) -> Result<()> {
+    let file = file_maybe_compressed(filepath, &opts).await?;
+    let file_size = file.metadata().await?.size();
```
Comment (adrian-kong) [inline crates/esthri/src/presign/upload.rs:66]:
```suggestion
    let file_size = file.metadata().await?.len();
```
should this be len() instead?

File: crates/esthri/Cargo.toml [hunk 3]
```diff
@@ -44,6 +44,8 @@ tokio-util = { version = "0.7", features = ["io"] }
 walkdir = "2"
 strum = "0.24.1"
 strum_macros = "0.24.3"
+reqwest = { version="0.11.12", features=["stream", "json"] }
```
Comment (adrian-kong) [inline crates/esthri/Cargo.toml:47]: could we slap this into test dependency,
presigned urls dont need https i believe

so u can probably also do
`default-features = false`

File: crates/esthri/src/presign/upload.rs (hunk 2 above)
Comment (john-michaelburke) [reply crates/esthri/src/presign/upload.rs:66]: Looks like they return the same thing :shrug: 
https://doc.rust-lang.org/std/fs/struct.Metadata.html#method.size

File: crates/esthri/src/presign/upload.rs (hunk 1 above)
Comment (joelynch) [reply crates/esthri/src/presign/upload.rs:13]: Good pick up, changed

File: crates/esthri/Cargo.toml (hunk 3 above)
Comment (joelynch) [reply crates/esthri/Cargo.toml:47]: reqwest is used in the presigned url client code - I could make this a specific feature maybe `presigned_url_client`.

**Comment by joelynch** [comment]
Maybe this work should go under some specific features

**Review by silverjam — APPROVED** [review]

**Pull request merged**